File: src/treeBurst/runtime/Scope.ts
import { GlobalScope, VOID } from "./GlobalScope"
import { ManagedValue } from "./ManagedValue"

export class Variable {
    public value: ManagedValue = VOID
}

export class Scope {
    public readonly variables = new Map<string, Variable>()

    public findVariable(name: string): Variable | null {
        return this.variables.get(name) ?? this.parent?.findVariable(name) ?? null
    }

    public declareVariable(name: string): Variable | null {
        if (this.variables.has(name)) {
            return null
        }

        const variable = new Variable()
        this.variables.set(name, variable)
        return variable
    }

    public makeChild() { return new Scope(this, this.globalScope) }

    constructor(
        public readonly parent: Scope | null,
        public readonly globalScope: GlobalScope,
    ) { }
}

File: packages/treeBurstJava/src/main/java/bt7s7k7/treeburst/runtime/Scope.java
package bt7s7k7.treeburst.runtime;

import java.util.HashMap;
import java.util.Map;

public class Scope {
	public final Map<String, Variable> variables = new HashMap<>();
	public final Scope parent;
	public final Realm realm;

	public Scope(Scope parent, Realm realm) {
		this.parent = parent;
		this.realm = realm;
	}

	public Variable findVariable(String name) {
		var variable = this.variables.get(name);
		if (variable != null) {
			return variable;
		}

		if (this.parent != null) {
			return this.parent.findVariable(name);
		}

		return null;
	}

	public Variable declareVariable(String name) {
		if (this.variables.containsKey(name)) {
			return null;
		}

		var variable = new Variable();
		this.variables.put(name, variable);
		return variable;
	}

	public Variable getOrDeclareLocal(String name) {
		return this.variables.computeIfAbsent(name, __ -> new Variable());
	}

	public Scope makeChild() {
		return new Scope(this, this.realm);
	}
}
